KAFKA-18855: Slice API for MemoryRecords

apoorvmittal10 opened the pull request:
The PR adds `slice` API in `Records.java` and further implementation in
`MemoryRecords`. With the addition of ShareFetch and it's support to
read from TieredStorage, where ShareFetch might acquire subset of fetch
batches and TieredStorage emits MemoryRecords, hence a slice API is
needed for MemoryRecords as well to limit the bytes transferred (if
subset batches are acquired).

MemoryRecords are sliced using `duplicate` and `slice` API of
ByteBuffer, which are backed by the original buffer itself hence no-copy
is created rather position, limit and offset are changed as per the new
position and length.

Reviewers: Andrew Schofield <aschofield@confluent.io>, Jun Rao
 <junrao@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

apoorvmittal10 committed:
KAFKA-18855: Slice API for MemoryRecords

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/common/record/Records.java:
```diff
@@ -90,4 +91,18 @@ public interface Records extends TransferableRecords {
+     * @param size The number of bytes after the start position to include
+     * @return A sliced wrapper on this message set limited based on the given position and size
+     */
+    Records slice(int position, int size) throws IOException;
```
This part doesn't seem right to me. Neither `MemoryRecords` nor `FileRecords` implements a slice method which returns `Records`. There would be a bit more work required to make the concrete classes properly interchangeable so the caller didn't need to know which class they were dealing with.

junrao replied:
Could we have both MemoryRecords and FileRecords return Records?

AndrewJSchofield replied:
Yes, we could, but there would be downstream effects of that.

apoorvmittal10 replied:
I have made a separate [commit](https://github.com/apache/kafka/pull/19581/commits/7eb661be549a31ba2339ee556cc313b96679c482) in this PR which shows the changes to make MemoryRecords and FileRecords to return Record type. The casting is required at 1 place in DumpLogSegments.scala. Couple of cast is required in test file where I am validating `validBytes` in MemoryRecordTest. Otherwise the change looks good.

junrao replied:
Thanks for he explanation. We can just keep what's in the PR then.

AndrewJSchofield reviewed (changes requested):
In general, looks good to me. However, I do have a doubt about the addition of the `slice()` method to the `Records` interface.

junrao commented on clients/src/main/java/org/apache/kafka/common/record/Records.java:
```diff
@@ -90,4 +91,18 @@ public interface Records extends TransferableRecords {
+     *
+     * If this records set is already sliced, the position will be taken relative to that slicing.
+     *
+     * @param position The start position to begin the read from
```
Could we add that position is expected to be aligned at the batch boundary? Otherwise, the returned Records can't be iterated.

apoorvmittal10 replied:
Done.

junrao commented on clients/src/main/java/org/apache/kafka/common/record/Records.java:
```diff
@@ -90,4 +91,18 @@ public interface Records extends TransferableRecords {
     Iterable<Record> records();
+
+    /**
+     * Return a slice of records from this instance, which is a view into this set starting from the given position
```
We are duplicating the same javadoc in 3 places. We could just keep it once in Records.

apoorvmittal10 replied:
Agree, done.

junrao commented on clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsTest.java:
```diff
@@ -1068,6 +1072,149 @@ public void testUnsupportedCompress() {
+    @ArgumentsSource(MemoryRecordsArgumentsProvider.class)
+    public void testSlice(Args args) {
+        // Create a MemoryRecords instance with multiple batches. Prior RecordBatch.MAGIC_VALUE_V2,
+        // every append in a batch is a new batch. After RecordBatch.MAGIC_VALUE_V2, we can have multiple
```
Hmm, even before MAGIC_VALUE_V2, a MemoryRecords can have multiple batches.

apoorvmittal10 replied:
So what I mean is that each `builder.append` in same `MemoryRecordsBuilder` creates new batch when magic is less than v2 and compression type is none. This can be seen in test case: https://github.com/apache/kafka/blob/7d027a4d83af15428d4aa9bdb2144adcad73408c/clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsTest.java#L1006. Do you think I need to add that when `compression type is none` in method comments for better clarity or I am missing something else?

junrao commented on clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsTest.java:
```diff
@@ -1068,6 +1072,149 @@ public void testUnsupportedCompress() {
+        assertEquals(remainingItems, batches(sliced), "Read starting from the third message");
+
+        // Read from second message and size is past the end of the file on the already sliced view.
+        sliced = records.slice(1, records.sizeInBytes() - 1)
```
The expected usage is that position will be aligned with batch boundary, right?

apoorvmittal10 replied:
Yes, that's true. And test case also verifies that though the first slice was random but subequent slice on same with correct batch boundary resulted in correct data.

junrao commented on clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsTest.java:
```diff
@@ -1068,6 +1072,149 @@ public void testUnsupportedCompress() {
+     */
+    @ParameterizedTest
+    @ArgumentsSource(MemoryRecordsArgumentsProvider.class)
+    public void testSliceForAlreadySlicedFileRecords(Args args) {
```
The name seems inaccurate since it's testing MemoryRecords.

apoorvmittal10 replied:
My bad, corrected.

junrao commented on clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsTest.java:
```diff
@@ -1068,6 +1072,149 @@ public void testUnsupportedCompress() {
+        assertEquals(records.sizeInBytes() - position, sliced.sizeInBytes());
+        assertEquals(items.subList(2, items.size()), batches(sliced), "Read starting from the third message");
+
+        // Further slice the already sliced memory records, from fourth message until the end. Now the
```
Not sure I understand the comment. The code slices from the 3rd message, instead of the 4th.

apoorvmittal10 replied:
So first we removed 2 batches at position 0 and 1, then we removed third batch at position 2. Hence I wrote, start reading from fourth till end.

junrao reviewed (commented):
@apoorvmittal10 : Thanks for the PR. Left a few comments.

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into KAFKA-18855

apoorvmittal10 committed:
Changing return type of slice API

apoorvmittal10 committed:
Addressing review comments

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into KAFKA-18855

apoorvmittal10 commented:
Thanks @AndrewJSchofield and @junrao for review, sorry I got occupied with other PR reviews and issues so couldn't reply earlier. I have addressed the comments now.

chia7712 commented on core/src/main/java/kafka/server/share/ShareFetchUtils.java:
```diff
@@ -209,16 +209,13 @@ static Partition partition(ReplicaManager replicaManager, TopicPartition tp) {
      *        of the fetched records. Otherwise, the original records are returned.
```
please update the docs.

apoorvmittal10 replied:
Thanks for the catch, done.

chia7712 commented on clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsTest.java:
```diff
@@ -1068,6 +1073,148 @@ public void testUnsupportedCompress() {
+            }
+            builder.close();
+        });
+        buffer.flip();
```
To be safer, especially if builder might create a new buffer for more spaces, it's better to use `builder#buffer` directly.

apoorvmittal10 replied:
So the buffer here is same for mutiple MemoryRecordsBuilder, to create different batches. Hence not used MemoryRecordsBuilder#buffer. Am I missing something?

chia7712 replied:
My point was `MemoryRecordsBuilder` could create another buffer if the input buffer is too small. However, I noticed that the test create a buffer with 1024 bytes so it should be fine

apoorvmittal10 replied:
Thanks @chia7712 for reviewing.

junrao commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java:
```diff
@@ -312,7 +312,7 @@ public ByteBuffer buffer() {
      */
     @Override
-    public MemoryRecords slice(int position, int size) {
+    public Records slice(int position, int size) {
```
Actually, it seems that our existing code already returns a subclass when implementing the Records interface.
For example, Records has 
`    Iterable<? extends RecordBatch> batches()`

and FileRecords has
`.   public Iterable<FileChannelRecordBatch> batches() `

Perhaps it's better to define the interface as 
`    <R extends Records> R slice(int position, int size) throws IOException`
and return `MemoryRecords` here. This way, we can avoid the type casting.

apoorvmittal10 replied:
Thanks for the suggestion @junrao . @AndrewJSchofield did also referenced same [approach](https://github.com/apache/kafka/blob/2df14b1190601955c01d4d74318093087601faa0/clients/src/main/java/org/apache/kafka/clients/admin/AbstractOptions.java#L25) in our discussion, but the issue I told him was that now where ever `Records` will be used then we have to specify the `type` else compilation will fail because of `warning: [rawtypes] usage` or we need to suppress warnings at all the places.

Second  option is currently in the PR, change the code to with minimal casting.

Third was the original, where I returned specific types from implementing classes. And Records just have Records as return type on slice API.

Fourth can be which I didn't implement is to not define `slice` API in `Records.java` and where slice API is needed and type is of `Records` then cast to the specific Records implementation type and use the `slice` API.

junrao commented on clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsTest.java:
```diff
@@ -1068,6 +1073,148 @@ public void testUnsupportedCompress() {
+
+        // Further slice the already sliced memory records, from fourth message until the end. Now the
+        // bytes available in the sliced records are less than the start position. However, the
+        // position to slice is relative hence reset position to second message in the sliced memory
```
> Now the bytes available in the sliced records are less than the start position.

Hmm, which start position does it refer to? Is that important?

apoorvmittal10 replied:
Yes, it's not important. Removed it.

junrao reviewed (commented):
@apoorvmittal10 : Thanks for the updated PR. A couple of more comments.

apoorvmittal10 committed:
Addressing review comments

junrao commented on clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsTest.java:
```diff
@@ -1068,6 +1073,146 @@ public void testUnsupportedCompress() {
+        assertEquals(items.subList(2, items.size()), batches(sliced), "Read starting from the third message");
+
+        // Further slice the already sliced memory records, from fourth message until the end. Now the
+        // bytes available in the sliced records are less than the start position.
```
I still don't quite understand this comment. Why are we comparing available bytes with the start position? Ditto in FileRecordsTest.

apoorvmittal10 replied:
So if a record has been sliced and now the available bytes is lesser than moved position i.e. the record has been sliced half past earlier then the slicing should work as well. This test was a result of failure we saw in FileRecords earlier and fixed in the this [PR](https://github.com/apache/kafka/pull/18759/files). I agree the comment here could be better and I am updating same.

junrao reviewed (commented):
@apoorvmittal10 : Thanks for the updated PR. One more comment.

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into KAFKA-18855

apoorvmittal10 committed:
Updating comment

junrao commented on clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsTest.java:
```diff
@@ -1068,6 +1073,146 @@ public void testUnsupportedCompress() {
+        assertEquals(items.subList(2, items.size()), batches(sliced), "Read starting from the third message");
+
+        // Further slice the already sliced memory records, from fourth message until the end. Now the
+        // bytes available in the sliced records are lesser than the moved position from original records.
```
lesser => less Ditto in FileRecordsTest.

apoorvmittal10 replied:
Done.

junrao reviewed (commented):
@apoorvmittal10 : Thanks for the updated PR. One more comment.

apoorvmittal10 committed:
Updating comment

junrao reviewed (approved):
@apoorvmittal10 : Thanks for the updated PR. LGTM assuming the tests pass.
